widgetssdk/src/main/java/com/glia/widgets/filepreview/data/GliaFileRepositoryImpl.kt

```diff
@@ -0,0 +1,104 @@
+        return Maybe.create { emitter: MaybeEmitter<Bitmap> ->
+            val bitmap = bitmapCache.getBitmapById(fileName)
+            if (bitmap != null) emitter.onSuccess(bitmap) else emitter.onError(
+                CacheFileNotFoundException()
+            )
+        }
+    }
+
+    override fun putImageToCache(fileName: String, bitmap: Bitmap): Completable {
+        return Completable.create { emitter: CompletableEmitter ->
+            try {
+                bitmapCache.putBitmap(fileName, bitmap)
+                emitter.onComplete()
+            } catch (ex: Exception) {
+                emitter.onError(ex)
+            }
+        }
+    }
+
+    override fun loadImageFromDownloads(fileName: String): Maybe<Bitmap> {
+        return downloadsFolderDataSource.getImageFromDownloadsFolder(fileName)
+    }
+
+    override fun putImageToDownloads(fileName: String, bitmap: Bitmap): Completable {
+        return downloadsFolderDataSource.putImageToDownloads(fileName, bitmap)
+    }
+
+    override fun loadImageFileFromNetwork(attachmentFile: AttachmentFile): Maybe<Bitmap> {
+        return Maybe.create { emitter: MaybeEmitter<InputStream?> ->
+            fetchFile(attachmentFile) { fileInputStream: InputStream, gliaException: GliaException? ->
+                if (gliaException != null) {
+                    emitter.onError(gliaException)
+                } else {
+                    emitter.onSuccess(fileInputStream)
+                }
+            }
+        }
+            .flatMap { inputStream: InputStream? ->
+                fileHelper.decodeSampledBitmapFromInputStream(inputStream)
+            }
+    }
+
+    override fun downloadFileFromNetwork(attachmentFile: AttachmentFile): Completable {
+        return Maybe.create { emitter: MaybeEmitter<InputStream> ->
+            fetchFile(attachmentFile) { fileInputStream: InputStream, gliaException: GliaException? ->
+                if (gliaException != null) {
+                    emitter.onError(gliaException)
+                } else {
+                    emitter.onSuccess(fileInputStream)
+                }
+            }
+        }
+            .flatMapCompletable { inputStream: InputStream ->
+                downloadsFolderDataSource
+                    .downloadFileToDownloads(
+                        FileHelper.getFileName(attachmentFile),
+                        attachmentFile.contentType,
+                        inputStream
+                    )
+                    .doOnComplete { inputStream.close() }
+            }
+    }
+
+    private fun fetchFile(attachmentFile: AttachmentFile, callback: RequestCallback<InputStream>) {
+        val engagement = gliaCore.currentEngagement.orElse(null)
+        if (engagement == null && engagementConfigRepository.chatType == ChatType.SECURE_MESSAGING) {
+            secureConversations.fetchFile(attachmentFile.id, callback)
+        } else {
+            gliaCore.fetchFile(attachmentFile, callback)
+        }
+    }
+}
```

Comment:
Note for future to help CR:
Make the change in the original java file, when feature has been reviewed and accepted, then translate to kotlin and retest before merger. This makes it better to review the specific changes only